### PR TITLE
fix(no-unused-vars): consider functions within conditional expressions usable

### PR DESCRIPTION
**Pull request by eventualbuddha**
Previously, a function declared directly within an `AstKind::ConditionalExpression` would be considered unused. This happened because `is_function_or_class_declaration_used` did not directly handle conditional expression parents, letting such cases fall into the default branch which just returned `false`, marking the function as unused. This change fixes that by considering such functions used when the conditional expression itself is used.

### CODE REVIEW

**Comment by DonIsaac** [comment]
Seems like a pathological case but otherwise LGTM

File: crates/oxc_linter/src/rules/eslint/no_unused_vars/tests/eslint.rs [hunk 1]
```diff
@@ -364,6 +364,9 @@ fn test() {
             None,
         ),
         ("function* foo(cb) { cb = yield function(a) { cb(1 + a); }; } foo();", None), // { "ecmaVersion": 6 },
+        ("console.log(function a() {} ? b : c)", None), // { "ecmaVersion": 6 }
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/no_unused_vars/tests/eslint.rs:367]: Can you add these to oxc test cases instead? These are the tests ported directly from ESLint

**Comment by eventualbuddha** [comment]
> Seems like a pathological case but otherwise LGTM

This actually came up as a false positive in my [company's codebase](https://github.com/votingworks/vxsuite/blob/326edf3e3d1c7776d2ef571a6811d8042a92c2eb/libs/eslint-plugin-vx/src/rules/gts_no_default_exports.ts#L102-L103). We name the generator function there because our ESLint config enables the [`func-names` rule](https://eslint.org/docs/latest/rules/func-names).

File: crates/oxc_linter/src/rules/eslint/no_unused_vars/tests/eslint.rs (hunk 1 above)
Comment (eventualbuddha) [reply crates/oxc_linter/src/rules/eslint/no_unused_vars/tests/eslint.rs]: a387d57

File: crates/oxc_linter/src/rules/eslint/no_unused_vars/tests/oxc.rs [hunk 2]
```diff
@@ -54,6 +54,9 @@ fn test_vars_simple() {
             ",
             None,
         ),
+        ("console.log(function a() {} ? b : c)", None),
```
Comment (camc314) [inline crates/oxc_linter/src/rules/eslint/no_unused_vars/tests/oxc.rs:57]: hopefully no one writes code like yhis. it'll always eval to b 🙂

**Review by camc314 — APPROVED** [review]

**Review by DonIsaac — APPROVED** [review]

**Pull request merged**